File: backend/src/main/java/com/relief/service/communication/DocumentCollaborationService.java
package com.relief.service.communication;

import com.relief.entity.User;
import com.relief.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

/**
 * Real-time document collaboration service
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class DocumentCollaborationService {

    private static final Logger log = LoggerFactory.getLogger(DocumentCollaborationService.class);
    private final UserRepository userRepository;
    private final Map<String, CollaborativeDocument> documents = new ConcurrentHashMap<>();
    private final Map<String, List<DocumentParticipant>> documentParticipants = new ConcurrentHashMap<>();
    private final Map<String, ReadWriteLock> documentLocks = new ConcurrentHashMap<>();
    private final Map<String, List<DocumentChange>> documentChanges = new ConcurrentHashMap<>();

    /**
     * Create a new collaborative document
     */
    @Transactional
    public CollaborativeDocument createDocument(String title, String content, UUID creatorId, DocumentType type) {
        log.info("Creating collaborative document: {} by user {}", title, creatorId);
        
        User creator = userRepository.findById(creatorId)
            .orElseThrow(() -> new IllegalArgumentException("Creator not found"));
        
        CollaborativeDocument document = new CollaborativeDocument();
        document.setId(UUID.randomUUID().toString());
        document.setTitle(title);
        document.setContent(content);
        document.setType(type);
        document.setCreatorId(creatorId);
        document.setCreatorName(creator.getFullName());
        document.setCreatedAt(LocalDateTime.now());
        document.setUpdatedAt(LocalDateTime.now());
        document.setStatus(DocumentStatus.ACTIVE);
        document.setVersion(1);
        document.setPermissions(DocumentPermissions.getDefaultPermissions());
        
        documents.put(document.getId(), document);
        documentParticipants.put(document.getId(), new ArrayList<>());
        documentChanges.put(document.getId(), new ArrayList<>());
        documentLocks.put(document.getId(), new ReentrantReadWriteLock());
        
        // Add creator as participant
        addParticipant(document.getId(), creatorId, ParticipantRole.OWNER);
        
        log.info("Collaborative document created: {} with ID: {}", document.getId(), document.getId());
        return document;
    }

    /**
     * Join a collaborative document
     */
    @Transactional
    public DocumentJoinResult joinDocument(String documentId, UUID userId) {
        log.info("User {} joining document {}", userId, documentId);
        
        CollaborativeDocument document = documents.get(documentId);
        if (document == null) {
            throw new IllegalArgumentException("Document not found");
        }
        
        // Check if user is already in document
        List<DocumentParticipant> participants = documentParticipants.get(documentId);
        boolean alreadyJoined = participants.stream()
            .anyMatch(p -> p.getUserId().equals(userId));
        
        if (alreadyJoined) {
            throw new IllegalStateException("User already in document");
        }
        
        // Add participant
        addParticipant(documentId, userId, ParticipantRole.COLLABORATOR);
        
        DocumentJoinResult result = new DocumentJoinResult();
        result.setDocument(document);
        result.setParticipant(getParticipant(documentId, userId));
        result.setWebSocketUrl(generateWebSocketUrl(documentId));
        result.setCollaborationConfig(generateCollaborationConfig(document));
        
        log.info("User {} successfully joined document {}", userId, documentId);
        return result;
    }

    /**
     * Apply changes to document
     */
    @Transactional
    public DocumentChangeResult applyChanges(String documentId, List<DocumentChange> changes, UUID userId) {
        log.info("Applying {} changes to document {} by user {}", changes.size(), documentId, userId);
        
        CollaborativeDocument document = documents.get(documentId);
        if (document == null) {
            throw new IllegalArgumentException("Document not found");
        }
        
        DocumentParticipant participant = getParticipant(documentId, userId);
        if (participant == null) {
            throw new IllegalStateException("User not in document");
        }
        
        ReadWriteLock lock = documentLocks.get(documentId);
        lock.writeLock().lock();
        
        try {
            List<DocumentChange> appliedChanges = new ArrayList<>();
            String newContent = document.getContent();
            
            for (DocumentChange change : changes) {
                if (isValidChange(change, newContent)) {
                    newContent = applyChange(newContent, change);
                    change.setAppliedAt(LocalDateTime.now());
                    change.setAppliedBy(userId);
                    appliedChanges.add(change);
                }
            }
            
            if (!appliedChanges.isEmpty()) {
                document.setContent(newContent);
                document.setUpdatedAt(LocalDateTime.now());
                document.setVersion(document.getVersion() + 1);
                
                // Store changes
                documentChanges.get(documentId).addAll(appliedChanges);
                
                // Update participant activity
                participant.setLastActivity(LocalDateTime.now());
                participant.setChangesCount(participant.getChangesCount() + appliedChanges.size());
            }
            
            DocumentChangeResult result = new DocumentChangeResult();
            result.setDocument(document);
            result.setAppliedChanges(appliedChanges);
            result.setVersion(document.getVersion());
            result.setSuccess(true);
            
            log.info("Applied {} changes to document {} (version: {})", appliedChanges.size(), documentId, document.getVersion());
            return result;
            
        } finally {
            lock.writeLock().unlock();
        }
    }

    /**
     * Get document with current content
     */
    @Transactional(readOnly = true)
    public CollaborativeDocument getDocument(String documentId) {
        CollaborativeDocument document = documents.get(documentId);
        if (document == null) {
            throw new IllegalArgumentException("Document not found");
        }
        return document;
    }

    /**
     * Get document participants
     */
    @Transactional(readOnly = true)
    public List<DocumentParticipant> getDocumentParticipants(String documentId) {
        return documentParticipants.getOrDefault(documentId, new ArrayList<>());
    }

    /**
     * Get document changes history
     */
    @Transactional(readOnly = true)
    public List<DocumentChange> getDocumentChanges(String documentId, int limit) {
        List<DocumentChange> changes = documentChanges.getOrDefault(documentId, new ArrayList<>());
        return changes.stream()
            .sorted(Comparator.comparing(DocumentChange::getAppliedAt).reversed())
            .limit(limit)
            .collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
    }

    /**
     * Update document permissions
     */
    @Transactional
    public CollaborativeDocument updatePermissions(String documentId, DocumentPermissions permissions, UUID userId) {
        CollaborativeDocument document = documents.get(documentId);
        if (document == null) {
            throw new IllegalArgumentException("Document not found");
        }
        
        DocumentParticipant participant = getParticipant(documentId, userId);
        if (participant == null || participant.getRole() != ParticipantRole.OWNER) {
            throw new IllegalStateException("Only document owner can update permissions");
        }
        
        document.setPermissions(permissions);
        document.setUpdatedAt(LocalDateTime.now());
        
        log.info("Document {} permissions updated by user {}", documentId, userId);
        return document;
    }

    /**
     * Leave document
     */
    @Transactional
    public void leaveDocument(String documentId, UUID userId) {
        log.info("User {} leaving document {}", userId, documentId);
        
        List<DocumentParticipant> participants = documentParticipants.get(documentId);
        if (participants != null) {
            participants.removeIf(p -> p.getUserId().equals(userId));
            
            // If no participants left, archive document
            if (participants.isEmpty()) {
                CollaborativeDocument document = documents.get(documentId);
                if (document != null) {
                    document.setStatus(DocumentStatus.ARCHIVED);
                    document.setUpdatedAt(LocalDateTime.now());
                }
            }
        }
        
        log.info("User {} left document {}", userId, documentId);
    }

    /**
     * Get user's documents
     */
    @Transactional(readOnly = true)
    public List<CollaborativeDocument> getUserDocuments(UUID userId) {
        return documents.values().stream()
            .filter(doc -> {
                List<DocumentParticipant> participants = documentParticipants.get(doc.getId());
                return participants != null && participants.stream()
                    .anyMatch(p -> p.getUserId().equals(userId));
            })
            .sorted(Comparator.comparing(CollaborativeDocument::getUpdatedAt).reversed())
            .collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
    }

    /**
     * Add participant to document
     */
    private void addParticipant(String documentId, UUID userId, ParticipantRole role) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new IllegalArgumentException("User not found"));
        
        DocumentParticipant participant = new DocumentParticipant();
        participant.setUserId(userId);
        participant.setUserName(user.getFullName());
        participant.setUserEmail(user.getEmail());
        participant.setRole(role);
        participant.setJoinedAt(LocalDateTime.now());
        participant.setLastActivity(LocalDateTime.now());
        participant.setChangesCount(0);
        
        documentParticipants.get(documentId).add(participant);
    }

    /**
     * Get participant from document
     */
    private DocumentParticipant getParticipant(String documentId, UUID userId) {
        List<DocumentParticipant> participants = documentParticipants.get(documentId);
        if (participants != null) {
            return participants.stream()
                .filter(p -> p.getUserId().equals(userId))
                .findFirst()
                .orElse(null);
        }
        return null;
    }

    /**
     * Check if change is valid
     */
    private boolean isValidChange(DocumentChange change, String content) {
        // Basic validation - in real implementation, use operational transformation
        return change.getPosition() >= 0 && change.getPosition() <= content.length();
    }

    /**
     * Apply change to content
     */
    private String applyChange(String content, DocumentChange change) {
        switch (change.getType()) {
            case INSERT:
                return content.substring(0, change.getPosition()) + 
                       change.getText() + 
                       content.substring(change.getPosition());
            case DELETE:
                int endPos = change.getPosition() + change.getLength();
                return content.substring(0, change.getPosition()) + 
                       content.substring(endPos);
            case REPLACE:
                int replaceEndPos = change.getPosition() + change.getLength();
                return content.substring(0, change.getPosition()) + 
                       change.getText() + 
                       content.substring(replaceEndPos);
            default:
                return content;
        }
    }

    /**
     * Generate WebSocket URL
     */
    private String generateWebSocketUrl(String documentId) {
        return "ws://relief-platform.com/ws/document/" + documentId;
    }

    /**
     * Generate collaboration config
     */
    private CollaborationConfig generateCollaborationConfig(CollaborativeDocument document) {
        CollaborationConfig config = new CollaborationConfig();
        config.setDocumentId(document.getId());
        config.setVersion(document.getVersion());
        config.setMaxParticipants(50);
        config.setRealTimeSync(true);
        config.setConflictResolution(true);
        config.setAutoSave(true);
        config.setAutoSaveInterval(30); // seconds
        return config;
    }

    // Data classes
    public static class CollaborativeDocument {
        private String id;
        private String title;
        private String content;
        private DocumentType type;
        private UUID creatorId;
        private String creatorName;
        private LocalDateTime createdAt;
        private LocalDateTime updatedAt;
        private DocumentStatus status;
        private int version;
        private DocumentPermissions permissions;

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getTitle() { return title; }
        public void setTitle(String title) { this.title = title; }

        public String getContent() { return content; }
        public void setContent(String content) { this.content = content; }

        public DocumentType getType() { return type; }
        public void setType(DocumentType type) { this.type = type; }

        public UUID getCreatorId() { return creatorId; }
        public void setCreatorId(UUID creatorId) { this.creatorId = creatorId; }

        public String getCreatorName() { return creatorName; }
        public void setCreatorName(String creatorName) { this.creatorName = creatorName; }

        public LocalDateTime getCreatedAt() { return createdAt; }
        public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

        public LocalDateTime getUpdatedAt() { return updatedAt; }
        public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

        public DocumentStatus getStatus() { return status; }
        public void setStatus(DocumentStatus status) { this.status = status; }

        public int getVersion() { return version; }
        public void setVersion(int version) { this.version = version; }

        public DocumentPermissions getPermissions() { return permissions; }
        public void setPermissions(DocumentPermissions permissions) { this.permissions = permissions; }
    }

    public static class DocumentParticipant {
        private UUID userId;
        private String userName;
        private String userEmail;
        private ParticipantRole role;
        private LocalDateTime joinedAt;
        private LocalDateTime lastActivity;
        private int changesCount;

        // Getters and setters
        public UUID getUserId() { return userId; }
        public void setUserId(UUID userId) { this.userId = userId; }

        public String getUserName() { return userName; }
        public void setUserName(String userName) { this.userName = userName; }

        public String getUserEmail() { return userEmail; }
        public void setUserEmail(String userEmail) { this.userEmail = userEmail; }

        public ParticipantRole getRole() { return role; }
        public void setRole(ParticipantRole role) { this.role = role; }

        public LocalDateTime getJoinedAt() { return joinedAt; }
        public void setJoinedAt(LocalDateTime joinedAt) { this.joinedAt = joinedAt; }

        public LocalDateTime getLastActivity() { return lastActivity; }
        public void setLastActivity(LocalDateTime lastActivity) { this.lastActivity = lastActivity; }

        public int getChangesCount() { return changesCount; }
        public void setChangesCount(int changesCount) { this.changesCount = changesCount; }
    }

    public static class DocumentChange {
        private String id;
        private String documentId;
        private ChangeType type;
        private int position;
        private int length;
        private String text;
        private UUID appliedBy;
        private LocalDateTime appliedAt;

        public DocumentChange() {
            this.id = UUID.randomUUID().toString();
        }

        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }

        public String getDocumentId() { return documentId; }
        public void setDocumentId(String documentId) { this.documentId = documentId; }

        public ChangeType getType() { return type; }
        public void setType(ChangeType type) { this.type = type; }

        public int getPosition() { return position; }
        public void setPosition(int position) { this.position = position; }

        public int getLength() { return length; }
        public void setLength(int length) { this.length = length; }

        public String getText() { return text; }
        public void setText(String text) { this.text = text; }

        public UUID getAppliedBy() { return appliedBy; }
        public void setAppliedBy(UUID appliedBy) { this.appliedBy = appliedBy; }

        public LocalDateTime getAppliedAt() { return appliedAt; }
        public void setAppliedAt(LocalDateTime appliedAt) { this.appliedAt = appliedAt; }
    }

    public static class DocumentJoinResult {
        private CollaborativeDocument document;
        private DocumentParticipant participant;
        private String webSocketUrl;
        private CollaborationConfig collaborationConfig;

        // Getters and setters
        public CollaborativeDocument getDocument() { return document; }
        public void setDocument(CollaborativeDocument document) { this.document = document; }

        public DocumentParticipant getParticipant() { return participant; }
        public void setParticipant(DocumentParticipant participant) { this.participant = participant; }

        public String getWebSocketUrl() { return webSocketUrl; }
        public void setWebSocketUrl(String webSocketUrl) { this.webSocketUrl = webSocketUrl; }

        public CollaborationConfig getCollaborationConfig() { return collaborationConfig; }
        public void setCollaborationConfig(CollaborationConfig collaborationConfig) { this.collaborationConfig = collaborationConfig; }
    }

    public static class DocumentChangeResult {
        private CollaborativeDocument document;
        private List<DocumentChange> appliedChanges;
        private int version;
        private boolean success;

        // Getters and setters
        public CollaborativeDocument getDocument() { return document; }
        public void setDocument(CollaborativeDocument document) { this.document = document; }

        public List<DocumentChange> getAppliedChanges() { return appliedChanges; }
        public void setAppliedChanges(List<DocumentChange> appliedChanges) { this.appliedChanges = appliedChanges; }

        public int getVersion() { return version; }
        public void setVersion(int version) { this.version = version; }

        public boolean isSuccess() { return success; }
        public void setSuccess(boolean success) { this.success = success; }
    }

    public static class DocumentPermissions {
        private boolean canEdit = true;
        private boolean canComment = true;
        private boolean canShare = true;
        private boolean canDelete = false;
        private boolean canManagePermissions = false;

        public static DocumentPermissions getDefaultPermissions() {
            return new DocumentPermissions();
        }

        // Getters and setters
        public boolean isCanEdit() { return canEdit; }
        public void setCanEdit(boolean canEdit) { this.canEdit = canEdit; }

        public boolean isCanComment() { return canComment; }
        public void setCanComment(boolean canComment) { this.canComment = canComment; }

        public boolean isCanShare() { return canShare; }
        public void setCanShare(boolean canShare) { this.canShare = canShare; }

        public boolean isCanDelete() { return canDelete; }
        public void setCanDelete(boolean canDelete) { this.canDelete = canDelete; }

        public boolean isCanManagePermissions() { return canManagePermissions; }
        public void setCanManagePermissions(boolean canManagePermissions) { this.canManagePermissions = canManagePermissions; }
    }

    public static class CollaborationConfig {
        private String documentId;
        private int version;
        private int maxParticipants;
        private boolean realTimeSync;
        private boolean conflictResolution;
        private boolean autoSave;
        private int autoSaveInterval; // seconds

        // Getters and setters
        public String getDocumentId() { return documentId; }
        public void setDocumentId(String documentId) { this.documentId = documentId; }

        public int getVersion() { return version; }
        public void setVersion(int version) { this.version = version; }

        public int getMaxParticipants() { return maxParticipants; }
        public void setMaxParticipants(int maxParticipants) { this.maxParticipants = maxParticipants; }

        public boolean isRealTimeSync() { return realTimeSync; }
        public void setRealTimeSync(boolean realTimeSync) { this.realTimeSync = realTimeSync; }

        public boolean isConflictResolution() { return conflictResolution; }
        public void setConflictResolution(boolean conflictResolution) { this.conflictResolution = conflictResolution; }

        public boolean isAutoSave() { return autoSave; }
        public void setAutoSave(boolean autoSave) { this.autoSave = autoSave; }

        public int getAutoSaveInterval() { return autoSaveInterval; }
        public void setAutoSaveInterval(int autoSaveInterval) { this.autoSaveInterval = autoSaveInterval; }
    }

    public enum DocumentType {
        EMERGENCY_PLAN, COORDINATION_DOC, TRAINING_MATERIAL, REPORT, MEETING_NOTES
    }

    public enum DocumentStatus {
        ACTIVE, ARCHIVED, DELETED
    }

    public enum ParticipantRole {
        OWNER, COLLABORATOR, VIEWER
    }

    public enum ChangeType {
        INSERT, DELETE, REPLACE
    }
}

File: frontend/src/services/documentCollaborationService.ts
/**
 * Document Collaboration Service
 * Handles real-time document collaboration and editing
 */

import { apiService } from './api';

export interface CollaborativeDocument {
  id: string;
  title: string;
  content: string;
  type: 'EMERGENCY_PLAN' | 'COORDINATION_DOC' | 'TRAINING_MATERIAL' | 'REPORT' | 'MEETING_NOTES';
  creatorId: string;
  creatorName: string;
  createdAt: string;
  updatedAt: string;
  status: 'ACTIVE' | 'ARCHIVED' | 'DELETED';
  version: number;
  permissions: DocumentPermissions;
}

export interface DocumentParticipant {
  userId: string;
  userName: string;
  userEmail: string;
  role: 'OWNER' | 'COLLABORATOR' | 'VIEWER';
  joinedAt: string;
  lastActivity: string;
  changesCount: number;
}

export interface DocumentChange {
  id: string;
  documentId: string;
  type: 'INSERT' | 'DELETE' | 'REPLACE';
  position: number;
  length: number;
  text: string;
  appliedBy: string;
  appliedAt: string;
}

export interface DocumentJoinResult {
  document: CollaborativeDocument;
  participant: DocumentParticipant;
  webSocketUrl: string;
  collaborationConfig: CollaborationConfig;
}

export interface DocumentChangeResult {
  document: CollaborativeDocument;
  appliedChanges: DocumentChange[];
  version: number;
  success: boolean;
}

export interface DocumentPermissions {
  canEdit: boolean;
  canComment: boolean;
  canShare: boolean;
  canDelete: boolean;
  canManagePermissions: boolean;
}

export interface CollaborationConfig {
  documentId: string;
  version: number;
  maxParticipants: number;
  realTimeSync: boolean;
  conflictResolution: boolean;
  autoSave: boolean;
  autoSaveInterval: number;
}

class DocumentCollaborationService {
  private activeDocuments: Map<string, CollaborativeDocument> = new Map();
  private documentParticipants: Map<string, DocumentParticipant[]> = new Map();
  private documentChanges: Map<string, DocumentChange[]> = new Map();
  private changeHandlers: Map<string, ((change: DocumentChange) => void)[]> = new Map();
  private participantHandlers: Map<string, ((participant: DocumentParticipant) => void)[]> = new Map();
  private autoSaveInterval: number | null = null;

  /**
   * Create a new collaborative document
   */
  async createDocument(title: string, content: string, type: string): Promise<CollaborativeDocument> {
    const response = await apiService.post('/document-collaboration/create', {
      title,
      content,
      type
    });

    this.activeDocuments.set(response.id, response);
    return response;
  }

  /**
   * Join a collaborative document
   */
  async joinDocument(documentId: string): Promise<DocumentJoinResult> {
    const response = await apiService.post(`/document-collaboration/${documentId}/join`);
    
    this.activeDocuments.set(documentId, response.document);
    this.documentParticipants.set(documentId, [response.participant]);
    
    return response;
  }

  /**
   * Apply changes to document
   */
  async applyChanges(documentId: string, changes: DocumentChange[]): Promise<DocumentChangeResult> {
    const response = await apiService.post(`/document-collaboration/${documentId}/changes`, {
      changes
    });

    if (response.success) {
      this.activeDocuments.set(documentId, response.document);
      
      // Store changes
      const existingChanges = this.documentChanges.get(documentId) || [];
      this.documentChanges.set(documentId, [...existingChanges, ...response.appliedChanges]);
      
      // Notify change handlers
      this.notifyChangeHandlers(documentId, response.appliedChanges);
    }

    return response;
  }

  /**
   * Get document details
   */
  async getDocument(documentId: string): Promise<CollaborativeDocument> {
    const response = await apiService.get(`/document-collaboration/${documentId}`);
    this.activeDocuments.set(documentId, response);
    return response;
  }

  /**
   * Get document participants
   */
  async getDocumentParticipants(documentId: string): Promise<DocumentParticipant[]> {
    const response = await apiService.get(`/document-collaboration/${documentId}/participants`);
    this.documentParticipants.set(documentId, response);
    return response;
  }

  /**
   * Get document changes history
   */
  async getDocumentChanges(documentId: string, limit: number = 50): Promise<DocumentChange[]> {
    const response = await apiService.get(`/document-collaboration/${documentId}/changes?limit=${limit}`);
    this.documentChanges.set(documentId, response);
    return response;
  }

  /**
   * Update document permissions
   */
  async updatePermissions(documentId: string, permissions: DocumentPermissions): Promise<CollaborativeDocument> {
    const response = await apiService.put(`/document-collaboration/${documentId}/permissions`, permissions);
    this.activeDocuments.set(documentId, response);
    return response;
  }

  /**
   * Leave document
   */
  async leaveDocument(documentId: string): Promise<void> {
    await apiService.post(`/document-collaboration/${documentId}/leave`);
    
    this.activeDocuments.delete(documentId);
    this.documentParticipants.delete(documentId);
    this.documentChanges.delete(documentId);
    this.changeHandlers.delete(documentId);
    this.participantHandlers.delete(documentId);
  }

  /**
   * Get user's documents
   */
  async getUserDocuments(): Promise<CollaborativeDocument[]> {
    const response = await apiService.get<CollaborativeDocument[]>('/document-collaboration/my-documents');
    return response;
  }

  /**
   * Get collaborative documents (alias for getUserDocuments for compatibility)
   */
  async getCollaborativeDocuments(): Promise<CollaborativeDocument[]> {
    return this.getUserDocuments();
  }

  /**
   * Create document change
   */
  createDocumentChange(
    documentId: string,
    type: 'INSERT' | 'DELETE' | 'REPLACE',
    position: number,
    length: number,
    text: string
  ): DocumentChange {
    return {
      id: `change_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      documentId,
      type,
      position,
      length,
      text,
      appliedBy: 'current_user',
      appliedAt: new Date().toISOString()
    };
  }

  /**
   * Add change handler
   */
  addChangeHandler(documentId: string, handler: (change: DocumentChange) => void): void {
    if (!this.changeHandlers.has(documentId)) {
      this.changeHandlers.set(documentId, []);
    }
    this.changeHandlers.get(documentId)!.push(handler);
  }

  /**
   * Remove change handler
   */
  removeChangeHandler(documentId: string, handler: (change: DocumentChange) => void): void {
    const handlers = this.changeHandlers.get(documentId);
    if (handlers) {
      const index = handlers.indexOf(handler);
      if (index > -1) {
        handlers.splice(index, 1);
      }
    }
  }

  /**
   * Add participant handler
   */
  addParticipantHandler(documentId: string, handler: (participant: DocumentParticipant) => void): void {
    if (!this.participantHandlers.has(documentId)) {
      this.participantHandlers.set(documentId, []);
    }
    this.participantHandlers.get(documentId)!.push(handler);
  }

  /**
   * Remove participant handler
   */
  removeParticipantHandler(documentId: string, handler: (participant: DocumentParticipant) => void): void {
    const handlers = this.participantHandlers.get(documentId);
    if (handlers) {
      const index = handlers.indexOf(handler);
      if (index > -1) {
        handlers.splice(index, 1);
      }
    }
  }

  /**
   * Start auto-save
   */
  startAutoSave(documentId: string, interval: number = 30000): void {
    this.stopAutoSave();
    
    this.autoSaveInterval = window.setInterval(() => {
      this.autoSaveDocument(documentId);
    }, interval);
  }

  /**
   * Stop auto-save
   */
  stopAutoSave(): void {
    if (this.autoSaveInterval) {
      clearInterval(this.autoSaveInterval);
      this.autoSaveInterval = null;
    }
  }

  /**
   * Auto-save document
   */
  private async autoSaveDocument(documentId: string): Promise<void> {
    try {
      const document = this.activeDocuments.get(documentId);
      if (document) {
        // In real implementation, save document content
        console.log('Auto-saving document:', documentId);
      }
    } catch (error) {
      console.error('Error auto-saving document:', error);
    }
  }

  /**
   * Notify change handlers
   */
  private notifyChangeHandlers(documentId: string, changes: DocumentChange[]): void {
    const handlers = this.changeHandlers.get(documentId);
    if (handlers) {
      changes.forEach(change => {
        handlers.forEach(handler => {
          try {
            handler(change);
          } catch (error) {
            console.error('Error in change handler:', error);
          }
        });
      });
    }
  }

  /**
   * Notify participant handlers
   */
  private notifyParticipantHandlers(documentId: string, participant: DocumentParticipant): void {
    const handlers = this.participantHandlers.get(documentId);
    if (handlers) {
      handlers.forEach(handler => {
        try {
          handler(participant);
        } catch (error) {
          console.error('Error in participant handler:', error);
        }
      });
    }
  }

  /**
   * Get active document
   */
  getActiveDocument(documentId: string): CollaborativeDocument | null {
    return this.activeDocuments.get(documentId) || null;
  }

  /**
   * Get cached document participants (synchronous, returns cached data)
   */
  getCachedDocumentParticipants(documentId: string): DocumentParticipant[] {
    return this.documentParticipants.get(documentId) || [];
  }

  /**
   * Get cached document changes (synchronous, returns cached data)
   */
  getCachedDocumentChanges(documentId: string): DocumentChange[] {
    return this.documentChanges.get(documentId) || [];
  }

  /**
   * Check if user can edit document
   */
  canEditDocument(documentId: string, userId: string): boolean {
    const document = this.activeDocuments.get(documentId);
    if (!document) return false;

    const participant = this.documentParticipants.get(documentId)?.find(p => p.userId === userId);
    if (!participant) return false;

    if (participant.role === 'OWNER') return true;
    if (participant.role === 'COLLABORATOR' && document.permissions.canEdit) return true;
    
    return false;
  }

  /**
   * Check if user can comment on document
   */
  canCommentOnDocument(documentId: string, userId: string): boolean {
    const document = this.activeDocuments.get(documentId);
    if (!document) return false;

    const participant = this.documentParticipants.get(documentId)?.find(p => p.userId === userId);
    if (!participant) return false;

    return document.permissions.canComment;
  }

  /**
   * Check if user can share document
   */
  canShareDocument(documentId: string, userId: string): boolean {
    const document = this.activeDocuments.get(documentId);
    if (!document) return false;

    const participant = this.documentParticipants.get(documentId)?.find(p => p.userId === userId);
    if (!participant) return false;

    return document.permissions.canShare;
  }

  /**
   * Get document type icon
   */
  getDocumentTypeIcon(type: string): string {
    switch (type) {
      case 'EMERGENCY_PLAN': return '🚨';
      case 'COORDINATION_DOC': return '📋';
      case 'TRAINING_MATERIAL': return '📚';
      case 'REPORT': return '📊';
      case 'MEETING_NOTES': return '📝';
      default: return '📄';
    }
  }

  /**
   * Get document status color
   */
  getDocumentStatusColor(status: string): string {
    switch (status) {
      case 'ACTIVE': return '#10b981'; // green
      case 'ARCHIVED': return '#6b7280'; // gray
      case 'DELETED': return '#ef4444'; // red
      default: return '#6b7280'; // gray
    }
  }

  /**
   * Get participant role icon
   */
  getParticipantRoleIcon(role: string): string {
    switch (role) {
      case 'OWNER': return '👑';
      case 'COLLABORATOR': return '✏️';
      case 'VIEWER': return '👁️';
      default: return '👤';
    }
  }

  /**
   * Get change type icon
   */
  getChangeTypeIcon(type: string): string {
    switch (type) {
      case 'INSERT': return '➕';
      case 'DELETE': return '➖';
      case 'REPLACE': return '🔄';
      default: return '📝';
    }
  }

  /**
   * Format document size
   */
  formatDocumentSize(content: string): string {
    const size = new Blob([content]).size;
    if (size < 1024) return `${size} B`;
    if (size < 1024 * 1024) return `${(size / 1024).toFixed(1)} KB`;
    return `${(size / (1024 * 1024)).toFixed(1)} MB`;
  }

  /**
   * Get document word count
   */
  getDocumentWordCount(content: string): number {
    return content.trim().split(/\s+/).filter(word => word.length > 0).length;
  }

  /**
   * Get document character count
   */
  getDocumentCharacterCount(content: string): number {
    return content.length;
  }

  /**
   * Get document line count
   */
  getDocumentLineCount(content: string): number {
    return content.split('\n').length;
  }

  /**
   * Get document reading time
   */
  getDocumentReadingTime(content: string): string {
    const wordCount = this.getDocumentWordCount(content);
    const readingTime = Math.ceil(wordCount / 200); // 200 words per minute
    return `${readingTime} min read`;
  }

  /**
   * Get document last activity
   */
  getDocumentLastActivity(document: CollaborativeDocument): string {
    const updatedAt = new Date(document.updatedAt);
    const now = new Date();
    const diff = now.getTime() - updatedAt.getTime();
    
    const minutes = Math.floor(diff / 60000);
    const hours = Math.floor(diff / 3600000);
    const days = Math.floor(diff / 86400000);
    
    if (minutes < 1) return 'Just now';
    if (minutes < 60) return `${minutes}m ago`;
    if (hours < 24) return `${hours}h ago`;
    return `${days}d ago`;
  }

  /**
   * Get participant last activity
   */
  getParticipantLastActivity(participant: DocumentParticipant): string {
    const lastActivity = new Date(participant.lastActivity);
    const now = new Date();
    const diff = now.getTime() - lastActivity.getTime();
    
    const minutes = Math.floor(diff / 60000);
    const hours = Math.floor(diff / 3600000);
    
    if (minutes < 1) return 'Active now';
    if (minutes < 60) return `${minutes}m ago`;
    if (hours < 24) return `${hours}h ago`;
    return 'Offline';
  }

  /**
   * Check if participant is active
   */
  isParticipantActive(participant: DocumentParticipant): boolean {
    const lastActivity = new Date(participant.lastActivity);
    const now = new Date();
    const diff = now.getTime() - lastActivity.getTime();
    
    return diff < 300000; // 5 minutes
  }

  /**
   * Get active participants count
   */
  getActiveParticipantsCount(documentId: string): number {
    const participants = this.documentParticipants.get(documentId) || [];
    return participants.filter(p => this.isParticipantActive(p)).length;
  }

  /**
   * Get total changes count
   */
  getTotalChangesCount(documentId: string): number {
    const changes = this.documentChanges.get(documentId) || [];
    return changes.length;
  }

  /**
   * Get changes count by user
   */
  getChangesCountByUser(documentId: string, userId: string): number {
    const changes = this.documentChanges.get(documentId) || [];
    return changes.filter(c => c.appliedBy === userId).length;
  }

  /**
   * Get document collaboration summary
   */
  getDocumentCollaborationSummary(documentId: string): {
    totalParticipants: number;
    activeParticipants: number;
    totalChanges: number;
    lastActivity: string;
  } {
    const participants = this.documentParticipants.get(documentId) || [];
    const changes = this.documentChanges.get(documentId) || [];
    const document = this.activeDocuments.get(documentId);
    
    return {
      totalParticipants: participants.length,
      activeParticipants: participants.filter(p => this.isParticipantActive(p)).length,
      totalChanges: changes.length,
      lastActivity: document ? this.getDocumentLastActivity(document) : 'Unknown'
    };
  }
}

export const documentCollaborationService = new DocumentCollaborationService();


